### PR TITLE
Added support for Amazon 2023, Debian 12, Ubuntu 24.04, removed EOL and BSD OSs and Python 2.7 support

### LINKED ISSUE #1940: [BUG] Minion bootstrap installation fails on Amazon Linux 2023
**Description**

Minion bootstrap installation fails on Amazon Linux 2023.  I initially discovered this problem when using 'salt-cloud' to spin up instances in AWS.  When that failed, I SSHed into the Amazon Linux minion and attempted to run the (newly downloaded) bootstrap script manually.

**Setup**
(Please provide relevant configs and/or SLS files (be sure to remove sensitive info. There is no general set-up of Salt.)

Please be as specific as possible and give set-up details.

- [ ] on-prem machine
- [ ] VM (Virtualbox, KVM, etc. please specify)
- [ X] VM running on a cloud service, please be explicit and add details
- [ ] container (Kubernetes, Docker, containerd, etc. please specify)
- [ ] or a combination, please be explicit
- [ ] jails if it is FreeBSD
- [ ] classic packaging
- [ ] onedir packaging
- [ ] used bootstrap to install

**Steps to Reproduce the behavior**
(Include debug logs if possible and relevant)

```
% ssh <AWS instance>

...
...

A newer release of "Amazon Linux" is available.
  Version 2023.0.20230614:
  Version 2023.1.20230628:
  Version 2023.1.20230629:
Run "/usr/bin/dnf check-release-update" for full release and version update info
   ,     #_
   ~\_  ####_        Amazon Linux 2023
  ~~  \_#####\
  ~~     \###|
  ~~       \#/ ___   https://aws.amazon.com/linux/amazon-linux-2023
   ~~       V~' '->
    ~~~         /
      ~~._.   _/
         _/ _/
       _/m/'
Last login: Fri Jun 30 13:53:03 2023 from 10.0.168.112
[ec2-user@ip-10-0-85-124 ~]$ sudo -i
[root@ip-10-0-85-124 ~]# curl -o bootstrap-salt.sh -L https://bootstrap.saltproject.io
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  345k  100  345k    0     0  1828k      0 --:--:-- --:--:-- --:--:-- 1836k
[root@ip-10-0-85-124 ~]# sh bootstrap-salt.sh
 *  INFO: Running version: 2023.04.26
 *  INFO: Executed by: sh
 *  INFO: Command line: 'bootstrap-salt.sh '
 *  WARN: Running the unstable version of bootstrap-salt.sh

 *  INFO: System Information:
 *  INFO:   CPU:          GenuineIntel
 *  INFO:   CPU Arch:     x86_64
 *  INFO:   OS Name:      Linux
 *  INFO:   OS Version:   6.1.29-47.49.amzn2023.x86_64
 *  INFO:   Distribution: Amazon Linux AMI 2023

 *  INFO: Installing minion
 *  INFO: Found function install_amazon_linux_ami_deps
 *  INFO: Found function config_salt
 *  INFO: Found function preseed_master
 *  INFO: Found function install_amazon_linux_ami_restart_daemons
 *  INFO: Found function daemons_running_onedir
 * ERROR: No installation function found. Exiting...
[root@ip-10-0-85-124 ~]#
```

**Expected behavior**

I expected to be able to install and configure the minion via the bootstrap method.

**Screenshots**

**Versions Report**
<details><summary>salt --versions-report</summary>
(Provided by running salt --versions-report. Please also mention any differences in master/minion versions.)

```yaml
PASTE HERE
```
</details>

**Additional context**

### PR DESCRIPTION
**Pull request by dmurphy18**
### What does this PR do?
Adds support for Amazon 2023, Debian 11 and 12.
Removes support for Debian 7, 8 & 9, RHEL 6, Amazon  Linux (RHEL 6 clone) and Python 2.7

### What issues does this PR fix or reference?
Fixes: https://github.com/saltstack/salt-bootstrap/issues/1940, https://github.com/saltstack/salt-bootstrap/issues/1986

### Previous Behavior
Amazon 2023, Debian 11 and 12 failed to install with bootstrap script

### New Behavior
Amazon 2023, Debian 11 and 12 now install with bootstrap script.
Debian 7, 8 and 9, RHEL 6, Amazon Linux (RHEL 6 clone) and Python 2.7 are no longer supported

### CODE REVIEW

**Comment by dmurphy18** [comment]
Cleaned up script with shellcheck before removing the EOL OS and Python 2.7 support.

File: bootstrap-salt.sh [hunk 1]
```diff
@@ -415,16 +416,16 @@ __usage() {
     -V  Install Salt into virtualenv
         (only available for Ubuntu based distributions)
     -x  Changes the Python version used to install Salt.
-        For CentOS 6 git installations python2.7 is supported.
-        Fedora git installation, CentOS 7, Ubuntu 18.04 support python3.
+        Python 2.7 is no longer supported.
+        Fedora git installation, CentOS 7, Ubuntu 1820.04 support python3.
```
Comment (twangboy) [inline bootstrap-salt.sh]: Ubuntu 1820?

File: bootstrap-salt.sh [hunk 2]
```diff
@@ -2079,7 +2088,7 @@ __wait_for_apt(){
     APT_RETURN=$?
 
     # Make sure we're not waiting on a lock
-    while [ $APT_RETURN -ne 0 ] && grep -q '^E: Could not get lock' "$APT_ERR"; do
+    while [ "$APT_RETURN" -ne 0 ] && grep -q '^E: Could not get lock' "$APT_ERR"; do
```
Comment (twangboy) [inline bootstrap-salt.sh:1929]: Don't you need to quote both sides of the comparison? `-ne "0"`
Does "0" == 0 in Linux?

File: bootstrap-salt.sh [hunk 3]
```diff
@@ -6529,110 +6605,240 @@ _eof
...
+
+        # This should prob be refactored to use __install_saltstack_rhel_repository()
+        # With args passed in to do the right thing.  Reformatted to be more like the
+        # amazon linux yum file.
+        if [ ! -s "/etc/yum.repos.d/${__REPO_FILENAME}" ]; then
+          cat <<_eof > "/etc/yum.repos.d/${__REPO_FILENAME}"
+[$repo_label]
+name=$repo_name
+failovermethod=priority
+priority=10
```
Comment (twangboy) [inline bootstrap-salt.sh]: Should these two be switched?

**Review by twangboy — DISMISSED** [review]
Do you mean to leave all the `## DGM` comments?

File: bootstrap-salt.sh (hunk 2 above)
Comment (dmurphy18) [reply bootstrap-salt.sh:1929]: yes, in bash
This is a work in progress

File: bootstrap-salt.sh (hunk 3 above)
Comment (dmurphy18) [reply bootstrap-salt.sh]: it is under heavy re-work, please wait till I remove the WIP

**Comment by dmurphy18** [comment]
Ensure the bootstrap script works on Centosstream 8 with redhat-lsb-core installed, it might be affecting distro

**Comment by Foorack** [comment]
CI indicates this is failing with `Failed to run install_debian_onedir_deps()!!!`, without any visible Error message.
Do we know why this is failing?

**Comment by dmurphy18** [comment]
@Foorack WIP - Work In Progress, read the title. Still have to implement GitHub Actions runners

**Comment by dmurphy18** [comment]
Note: nightly workflow directory structure has changed and needs adjustment in the bootstrap script, for example: https://repo.saltproject.io/salt-dev/3006.x/latest/salt/py3/photon/4.0/x86_64/ etc. and that **_latest_** could be replaced by a date, that is, https://repo.saltproject.io/salt-dev/3006.x/2024-07-02/salt/py3/photon/4.0/x86_64/minor/

**Comment by dmurphy18** [comment]
Having issues with Centos Stream 8
```
david@david-XPS-15-9570:~$ docker run -it --rm --net=host quay.io/centos/centos:stream8
Unable to find image 'quay.io/centos/centos:stream8' locally
stream8: Pulling from centos/centos
3ca969c7b6d7: Pull complete 
Digest: sha256:20da069d4f8126c4517ee563e6e723d4cbe79ff62f6c4597f753478af91a09a3
Status: Downloaded newer image for quay.io/centos/centos:stream8
[root@david-XPS-15-9570 /]# yum update
Failed to set locale, defaulting to C.UTF-8
CentOS Stream 8 - AppStream                                                                                                                                                                                                                                                                       0.0  B/s |   0  B     00:02    
Errors during downloading metadata for repository 'appstream':
  - Curl error (6): Couldn't resolve host name for http://mirrorlist.centos.org/?release=8-stream&arch=x86_64&repo=AppStream&infra=container [Could not resolve host: mirrorlist.centos.org]
Error: Failed to download metadata for repo 'appstream': Cannot prepare internal mirrorlist: Curl error (6): Couldn't resolve host name for http://mirrorlist.centos.org/?release=8-stream&arch=x86_64&repo=AppStream&infra=container [Could not resolve host: mirrorlist.centos.org]
[root@david-XPS-15-9570 /]# exit
```

Cannot even get a docker container locally, and then this https://blog.centos.org/2023/04/end-dates-are-coming-for-centos-stream-8-and-centos-linux-7/ so Centos Stream 8 is EOL as of Mat 31, 2024

Wondering if should just remove Centos Stream 8 since EOL

**Comment by dmurphy18** [comment]
Photon OS 4 has git install issues with the compiler
```
2024-07-03T21:29:35.7709169Z          self.initialize_options()
2024-07-03T21:29:37.9135494Z        /usr/lib/python3.10/site-packages/setuptools/_distutils/cmd.py:66: SetuptoolsDeprecationWarning: setup.py install is deprecated.
2024-07-03T21:29:37.9136909Z        !!
2024-07-03T21:29:37.9137295Z        
2024-07-03T21:29:37.9137763Z         ********************************************************************************
2024-07-03T21:29:37.9138529Z         Please avoid running ``setup.py`` directly.
2024-07-03T21:29:37.9139530Z         Instead, use pypa/build, pypa/installer or other
2024-07-03T21:29:37.9140329Z         standards-based tools.
2024-07-03T21:29:37.9140848Z        
2024-07-03T21:29:37.9141775Z         See https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html for details.
2024-07-03T21:29:37.9142837Z         ********************************************************************************
2024-07-03T21:29:37.9143478Z        
2024-07-03T21:29:37.9143818Z        !!
2024-07-03T21:29:37.9144217Z          self.initialize_options()
2024-07-03T21:29:37.9803124Z        x86_64-unknown-linux-gnu-gcc: fatal error: cannot execute 'as': execvp: No such file or directory
2024-07-03T21:29:37.9804014Z        compilation terminated.
2024-07-03T21:29:37.9819795Z        error: Setup script exited with error: command '/usr/bin/x86_64-unknown-linux-gnu-gcc' failed with exit code 1
2024-07-03T21:29:38.0412963Z        + return 1
2024-07-03T21:29:38.0415365Z        + echoerror 'Failed to run install_photon_git()!!!'
2024-07-03T21:29:38.0417639Z        + printf '\033[1;31m * ERROR\033[0m: %s\n' 'Failed to run install_photon_git()!!!'
2024-07-03T21:29:38.0419521Z        [1;31m * ERROR[0m: Failed to run install_photon_git()!!!
```

**Comment by dmurphy18** [comment]
Running into issue with Amazon 2023 with current stable version Ruby 3.3.4 for Create Test Instance, reverting to original 2.6.10
```
11s
Run ## DGM Need to replace this
-----> Starting Test Kitchen (v3.6.0)
-----> Creating <stable-3006-amazon-2023>...
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: [1](https://github.com/saltstack/salt-bootstrap/actions/runs/9882490882/job/27295489413?pr=1987#step:6:1) actions failed.
>>>>>>     Create failed on instance <stable-3006-amazon-2023>.  Please see .kitchen/logs/stable-3006-amazon-2023.log for more details
>>>>>> ----------------------
>>>>>> Please see .kitchen/logs/kitchen.log for more details
>>>>>> Also try running `kitchen diagnose --all` for configuration
Error: Process completed with exit code 20.
```

File: bootstrap-salt.sh (hunk 2 above)
Comment (twangboy) [reply bootstrap-salt.sh:1929]: Still unquoted

**Review by twangboy — APPROVED** [review]

**Pull request merged**